sdk/lib/atl/atlconv.h

```diff
@@ -0,0 +1,312 @@
+        Init(psz);
+        UNREFERENCED_PARAMETER(nCodePage);
+    }
+
+    ~CW2WEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPWSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CW2WEX is not copyable
+    CW2WEX(_In_ const CW2WEX&) throw();
+    CW2WEX& operator=(_In_ const CW2WEX&) throw();
+
+    void Init(_In_z_ LPCWSTR psz)
+    {
+        int cchMax = lstrlenW(psz);
+        if (cchMax <= t_nBufferLength)
+        {
+#ifdef _STRSAFE_H_INCLUDED_
+            StringCchCopyW(m_szBuffer, _countof(m_szBuffer), psz);
+#else
+            lstrcpynW(m_szBuffer, psz, _countof(m_szBuffer));
+#endif
+            m_psz = m_szBuffer;
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = _wcsdup(psz);
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CA2WEX
+{
+public:
+    LPWSTR m_psz;
+    wchar_t m_szBuffer[t_nBufferLength];
+
+    CA2WEX(_In_z_ LPCSTR psz)
+    {
+        Init(psz, CP_ACP);
+    }
+
+    CA2WEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz, nCodePage);
+    }
+
+    ~CA2WEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPWSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CA2WEX is not copyable
+    CA2WEX(_In_ const CA2WEX&) throw();
+    CA2WEX& operator=(_In_ const CA2WEX&) throw();
+
+    void Init(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        int cchMax = MultiByteToWideChar(nCodePage, 0, psz, -1, NULL, 0);
+        if (cchMax <= t_nBufferLength)
+        {
+            MultiByteToWideChar(nCodePage, 0, psz, -1, m_szBuffer, _countof(m_szBuffer));
+            m_szBuffer[_countof(m_szBuffer) - 1] = 0;
+            m_psz = m_szBuffer;
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = (LPWSTR)malloc(cchMax * sizeof(WCHAR));
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+
+        MultiByteToWideChar(nCodePage, 0, psz, -1, m_psz, cchMax);
+        m_psz[cchMax - 1] = 0;
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CW2AEX
+{
+public:
+    LPSTR m_psz;
+    char m_szBuffer[t_nBufferLength];
+
+    CW2AEX(_In_z_ LPCWSTR psz)
+    {
+        Init(psz, CP_ACP);
+    }
+
+    CW2AEX(_In_z_ LPCWSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz, nCodePage);
+    }
+
+    ~CW2AEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CW2AEX is not copyable
+    CW2AEX(_In_ const CW2AEX&) throw();
+    CW2AEX& operator=(_In_ const CW2AEX&) throw();
+
+    void Init(_In_z_ LPCWSTR psz, _In_ UINT nConvertCodePage)
+    {
+        int cchMax = WideCharToMultiByte(nConvertCodePage, 0, psz, -1, NULL, 0, NULL, NULL);
+        if (cchMax <= t_nBufferLength)
```

Comment:
(Note that `_countof(m_szBuffer)` equals `t_nBufferLength`)

Reply:
Fixed.